src/client/mod.rs

```diff
@@ -1,322 +1,498 @@
 //! This module provides the Warpgrapher client.
 
-use super::error::{Error, ErrorKind};
+use crate::engine::context::{GlobalContext, RequestContext};
+use crate::{Engine, Error, GraphQLRequest};
 use inflector::Inflector;
-use log::debug;
-use serde_json::{json, Value};
-use std::collections::BTreeMap;
+use log::{debug, trace};
+use serde_json::{from_value, json, Value};
+use std::collections::HashMap;
+use std::fmt::Display;
+use std::sync::mpsc;
+use std::thread;
 
 /// A Warpgrapher GraphQL client
 ///
 /// The [`Client`] provides a set of CRUD operations that will
 /// automatically generate GraphQL queries that conform to the wargrapher API
 ///
-/// [`Client`]: ./struct.Client.html
+/// [`Client`]: ./enum.Client.html
 ///
 /// # Examples
 ///
 /// ```rust
-/// use std::env::var_os;
-/// use warpgrapher::client::Client;;
+/// # use warpgrapher::Client;;
 ///
-/// let mut client = Client::new("http://localhost:5000/graphql");
+/// let client = Client::<(), ()>::new_with_http("http://localhost:5000/graphql");
 /// ```
-#[derive(Clone, Hash, Debug, Default)]
-pub struct Client {
-    endpoint: String,
+#[derive(Clone, Debug)]
+pub enum Client<GlobalCtx = (), RequestCtx = ()>
+where
+    GlobalCtx: GlobalContext,
+    RequestCtx: RequestContext,
+{
+    Http {
+        endpoint: String,
+    },
+    Local {
+        engine: Box<Engine<GlobalCtx, RequestCtx>>,
+    },
 }
 
-impl Client {
-    /// Takes an endpoint string and creates a new [`Client`].
+impl<GlobalCtx, RequestCtx> Client<GlobalCtx, RequestCtx>
+where
+    GlobalCtx: GlobalContext,
+    RequestCtx: RequestContext,
+{
+    /// Takes the URL of a Warpgrapher service endpoint and returns a new ['Client'] initialized to
+    /// query that endpoint.  The type parameters are only relevant for a local instance of the
+    /// Warpgrapher engine, not for a remote HTTP client, so pass () for both type parameters, as
+    /// shown in the example below.
+    ///
+    /// [`Client`]: ./enum.Client.html
     ///
-    /// [`Client`]: ./struct.Client.html
     ///
     /// # Examples
     ///
     /// ```rust
-    /// use std::env::var_os;
-    /// use warpgrapher::client::Client;
+    /// # use warpgrapher::Client;
     ///
-    /// let mut client = Client::new("http://localhost:5000/graphql");
+    /// let mut client = Client::<(), ()>::new_with_http("http://localhost:5000/graphql");
```

Comment:
@maldrake : Does rust have any way to specify default type parameters so as to not need to specify two empty objects?

Reply:
Yes, it is possible.  There are a handful of places where the interaction, or lack thereof, between default type parameters and type inference are described, but the one at [https://users.rust-lang.org/t/default-trait-type-not-working-for-type-inference/33905/2](https://users.rust-lang.org/t/default-trait-type-not-working-for-type-inference/33905/2) was the clearest and most succinct that I've found.

However, the syntax for the workaround isn't great, either.  It would be `<Client>::new_with_http` rather than `Client::<(), ()>`.  It's a little more concise, but I suspect most novice and even some experience Rust developers would not have encountered that syntax before.

Another option would be to introduce a trait in the middle somewhere, capturing the type parameters to Engine as associated types.  However, that would force the user to create a struct implementing that trait, specifying the actuals for the associated types, and implementing at minimum a get/set pair for an engine, which seemed even clunkier.